### PR TITLE
Implement listing images/containers using React JSX

### PR DESCRIPTION
**Pull request by QiWang19**
Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by QiWang19** [comment]
@martinpitt PTAL

File: src/ContainerDetails.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import cockpit from 'cockpit';
+
+const _ = cockpit.gettext;
+class ContainerDetails extends React.Component {
+    constructor(props) {
+        super(props);
+    }
```
Comment (martinpitt) [inline src/ContainerDetails.js]: This component has no state, so it should be a  [functional component](https://reactjs.org/docs/components-and-props.html#functional-and-class-components). ([another interesting blog post](https://hackernoon.com/react-stateless-functional-components-nine-wins-you-might-have-overlooked-997b0d933dbc)). Also, it's custom to explicitly spell out the properties that you support/expect, like this:
```js
const ContainerDetails({container}) => (
    <div> ...
        <dt>{ container.Created.. }</dt>
    </div>
);
```

File: package.json [hunk 2]
```diff
@@ -11,29 +11,45 @@
...
     "sizzle": "^2.3.3",
     "stdio": "^0.2.7",
     "webpack": "^2.6.1"
   },
   "dependencies": {
+    "bootstrap": "^4.1.1",
+    "bootstrap-less": "^3.3.8",
+    "create-react-class": "^15.6.3",
+    "exenv": "^1.2.2",
+    "jquery": "^3.3.1",
```
Comment (martinpitt) [inline package.json]: Please avoid introducing jQuery in new code, and particularly not with react. I don't see it actually being used here? I think this can just be dropped. Likewise, please comb the list for other unused stuff, like "exenv" -- it just unnecessarily enlarges `npm install` time, and all of these dependencies introduce maintenance. `react-create-class` also sounds like dead weight -- use proper ES6 classes instead.

File: src/ContainerDetails.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+
+    render() {
+        let container = this.props.container;
+        return (
+            <div className='listing-ct-body'>
+                <dl>
+                    <dt>{_("Id")}      </dt> <dd>{ container.ID }</dd>
```
Comment (martinpitt) [inline src/ContainerDetails.js]: Please avoid the extra whitespace in the element. Perhaps they are not rendered, but it might still cause some layout issues. Also please be consistent with putting the `<dd>`s into a new line.

"Id" should be spelled "ID", as it's an acronym.

File: src/Containers.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import Listing from '../lib/cockpit-components-listing.jsx';
+import cockpit from 'cockpit';
+
+import ContainerDetails from './ContainerDetails.js';
+const _ = cockpit.gettext;
+
+class Containers extends React.Component {
+    constructor(props) {
+	super(props);
+	console.log(props.containers);
```
Comment (martinpitt) [inline src/Containers.js]: Debugging leftover.

File: src/Containers.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    //TODO
+    restartContainer(container) {
+	return undefined;
+    }
+
+    getContainerState(state) {
+	if (state.Status)
+	    return state.Status;
+	if (state.Running)
+	    return "running";
```
Comment (martinpitt) [inline src/Containers.js]: Needs i18n (`_("")` wrapper)

File: src/Containers.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    //TODO
+    restartContainer(container) {
+	return undefined;
+    }
+
+    getContainerState(state) {
+	if (state.Status)
+	    return state.Status;
```
Comment (martinpitt) [inline src/Containers.js]: This is confusing. Why is it sometimes a free-form `Status` and sometimes a fixed `state.Paused`? This deserves a comment, or only using one of them if it is redundant.

File: src/Containers.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	if (state.Running)
+	    return "running";
+	if (state.Paused)
+	    return "paused";
+	if (state.Restarting)
+	    return "restarting";
+	if (state.FinishedAt && state.FinishedAt.indexOf("0001") === 0)
```
Comment (martinpitt) [inline src/Containers.js]: This "magic" string and comparison needs a comment.

File: src/Containers.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	    return "created";
+	return "exited";
+    }
```
Comment (martinpitt) [inline src/Containers.js]: missing empty line between methods (same error several times above).

File: src/Containers.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    renderRow(container) {
+	let isRunning = !!container.State.Running;
+	//TODO: check container.Image == container.ImageID
+	let image = container.ImageName;
+	let state = this.getContainerState(container.State);
```
Comment (martinpitt) [inline src/Containers.js]: supposedly these should be  `const`

File: src/Containers.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+	let columns = [
+	    { name: container.Name.replace(/^\//, ''), header: true },
```
Comment (martinpitt) [inline src/Containers.js]: What's wrong with showing slashes in the name?

File: src/Containers.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	    image,
+	    // util.render_container_cmdline(container),
+	    container.Config.Entrypoint === "" ? container.Config.Cmd.join(" "):container.Config.Cmd.join(""),
```
Comment (martinpitt) [inline src/Containers.js]: same value in both cases, and missing spaces around operators. You probably should enable eslint for this project. (And I'll look into doing that for the starter kit)

File: src/Containers.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	];
+	let tabs = [{
+	    name: _("Details"),
+	    renderer: ContainerDetails,
+	    data: { container: container }
+	}];
+
+	let startStopActions = [];
+	if (isRunning)
+	    startStopActions.push({ label: _("Stop"), onActivate: this.stopContainer.bind(this, container) });
```
Comment (martinpitt) [inline src/Containers.js]: Don't bind methods in a `render` function. It's a more common (and efficient?) style to write something like

    onActivate: () => this.stopContainer(container)

File: src/Containers.js [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	else
+	    startStopActions.push({ label: _("Start"), onActivate: this.startContainer.bind(this, container) });
+
+	startStopActions.push({
+	    label: _("Restart"),
+	    onActivate: this.restartContainer.bind(this, container),
+	    disabled: !isRunning
```
Comment (martinpitt) [inline src/Containers.js]: Why bother adding a disabled menu entry?

File: src/Containers.js [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+	return <Listing.ListingRow key={container.Id}
+				   columns={columns}
+				   tabRenderers={tabs}
+				   navigateToItem={ this.navigateToContainer.bind(this, container) }
+				   listingActions={actions} />;
+    }
+
+    render() {
+	console.log(this.props.onlyShowRunning);
```
Comment (martinpitt) [inline src/Containers.js]: debugging leftover

File: src/Containers.js [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+	const _ = cockpit.gettext;
+	const columnTitles = [_("Name"), _("Image"), _("Command"), _("CPU"), _("Memory"), _("State")];
+	//TODO
+	let emptyCaption = _("No containers");
+
+	let filtered = this.props.containers.filter((container)=>{
```
Comment (martinpitt) [inline src/Containers.js]: Missing spaces around operator, and unnecessary braces around a single argument.

File: src/Containers.js [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+	    if (this.props.onlyShowRunning && !container.State.Running)
+		return false;
+	    return true;
```
Comment (martinpitt) [inline src/Containers.js]: write this more succinctly as `filter(container => (!this.props.onlyShowRunning || container.State.Running));

File: src/ContainersRunImageModal.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import Modal from 'react-modal';
+
+class ContainersRunImageModal extends React.Component {
+    constructor(props) {
+      super(props);
+
+    }
```
Comment (martinpitt) [inline src/ContainersRunImageModal.js]: Unnecessary if you don't do anything other than calling the parent. Also, this component can be functional, it doesn't have any state.

File: src/ContainersRunImageModal.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+
+    render() {
+      return (
+	<Modal
+	  isOpen={this.props.show}
+	>
```
Comment (martinpitt) [inline src/ContainersRunImageModal.js]: IMHO this looks really ugly - just write it as one line, it's short enough.

File: src/Containers.js (hunk 10 above)
Comment (QiWang19) [reply src/Containers.js]: copied from dockerUI containers-view. Maybe deal with the different version of dockerAPI?

File: src/Containers.js (hunk 11 above)
Comment (QiWang19) [reply src/Containers.js]: Not the same. Add a space in the first case. The other is to join the directory.

File: src/Containers.js (hunk 13 above)
Comment (QiWang19) [reply src/Containers.js]: Haven't implement the function yet. Just want it can be run correctly.

File: src/Containers.js (hunk 6 above)
Comment (QiWang19) [reply src/Containers.js]: Copied from dockerUI. Unnecessary for varlinkAPI, I've deleted it.

File: package.json (hunk 2 above)
Comment (QiWang19) [reply package.json]: I changed it. But dockerUI uses Id.

**Comment by QiWang19** [comment]
I fixed the review issues you found. @martinpitt PTAL

File: src/index.html [hunk 19]
```diff
@@ -22,7 +22,10 @@
     <meta name="description" content="">
     <meta name="viewport" content="width=device-width, initial-scale=1">
 
+
+
     <link rel="stylesheet" href="../base1/patternfly.css">
+    <link href="docker.css" type="text/css" rel="stylesheet" id="term-style">
```
Comment (martinpitt) [inline src/index.html]: Please rename this to podman.css (same for the file) to avoid confusion.

File: src/Containers.js [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+	let columns = [
+		{ name: container.Name, header: true },
+		image,
+		container.Config.Entrypoint === "" ? container.Config.Cmd.join(" ") : container.Config.Cmd.join(""),
+		// TODO: CpuUsage
+		"",
+		// TODO: MemoryUsage, MemoryLimit
+		"",
+		state,
```
Comment (martinpitt) [inline src/Containers.js]: This should still be i18n'ed, but I'm happy for this to be a `// TODO: i18n` comment for now.

File: src/Containers.js [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+const renderRow = (containersStats, container) => {
+	const isRunning = !!container.State.Running;
+	//TODO: check container.Image == container.ImageID
+	const image = container.ImageName;
+	const state = container.State.Status;
+
+	let columns = [
+		{ name: container.Name, header: true },
+		image,
+		container.Config.Entrypoint === "" ? container.Config.Cmd.join(" ") : container.Config.Cmd.join(""),
```
Comment (martinpitt) [inline src/Containers.js]: The second join is still confusing, please add a comment what it does.

File: webpack.config.js [hunk 22]
```diff
@@ -156,6 +156,35 @@ module.exports = {
...
+                loader: 'strict-loader' // Adds "use strict"
+            },
+            {
+                test: /\.jsx$/,
+                loader: "babel-loader"
+            },
+            {
+                test: /\.es6$/,
+                loader: "babel-loader"
+            },
```
Comment (martinpitt) [inline webpack.config.js]: These all already exist above and are duplicated.

File: src/ContainerDetails.js [hunk 23]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
```
Comment (martinpitt) [inline src/ContainerDetails.js]: This file must be named .jsx, to apply correct babel and eslint rules. It's not "classic" non-ES6 JavaScript. Same applies to the other *.js files in this PR.

File: webpack.config.js [hunk 24]
```diff
@@ -156,6 +156,35 @@ module.exports = {
...
+            {
+                test: /\.css$/,
+                loader: extract.extract("css-loader?minimize=&root=" + libdir)
+            },
```
Comment (martinpitt) [inline webpack.config.js]: You can also rename your .css to .scss so that over time it can uses the newer "S"css features.

File: webpack.config.js [hunk 25]
```diff
@@ -156,6 +156,35 @@ module.exports = {
...
+            {
+                test: /\.less$/,
+                loader: extract.extract("css-loader?sourceMap&minimize=!less-loader?sourceMap&compress=false")
+            },
+            {
+                test: /views\/[^\/]+\.html$/,
+                loader: "ng-cache?prefix=[dir]"
+            },
```
Comment (martinpitt) [inline webpack.config.js]: This looks unrelated and should be a separate commit with a proper justification.

File: webpack.config.js [hunk 26]
```diff
@@ -156,6 +156,35 @@ module.exports = {
...
+            {
+                test: /[\/]angular\.js$/,
+                loader: "exports?angular"
             }
         ]
```
Comment (martinpitt) [inline webpack.config.js]: There's no angular in this code, and you should stay really far away from introducing it :-) Please drop this.

File: src/ContainerHeader.js [hunk 27]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import cockpit from 'cockpit';
+import Select from '../lib/cockpit-components-select.jsx';
+
+const _ = cockpit.gettext;cockpit.gett
```
Comment (martinpitt) [inline src/ContainerHeader.js]: There's some junk here

File: src/ContainerHeader.js [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+
+class ContainerHeader extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            filter: 'running',
+            filterText: ''
+        };
+
+        this.filterChanged = this.filterChanged.bind(this);
```
Comment (martinpitt) [inline src/ContainerHeader.js]: This doesn't seem necessary, you only call this from a method that already has a correct `this`.

File: src/ContainerHeader.js [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    filterChanged() {
+        if (this.props.onFilterChanged)
+            this.props.onFilterChanged(this.state.filter, this.state.filterText);
+    }
+
+    handleFilterChange (value) {
+        this.setState({ filter: value });
+        this.props.onChange(value);
```
Comment (martinpitt) [inline src/ContainerHeader.js]: Should check if the `onChange` property is actually defined. Or declare it a mandatory property.

**Comment by QiWang19** [comment]
@martinpitt I have made the changes you requested.  Sorry, I am still getting used to the github workflow.

File: webpack.config.js [hunk 30]
```diff
@@ -156,7 +156,24 @@ module.exports = {
...
+                loader: "babel-loader"
+            },
+            {
+                test: /\.css$/,
+                loader: extract.extract("css-loader?minimize=&root=" + libdir)
+            },
+            {
+                test: /\.less$/,
+                loader: extract.extract("css-loader?sourceMap&minimize=!less-loader?sourceMap&compress=false")
+            },
```
[the remark this answers is not in the document]
Comment (QiWang19) [reply webpack.config.js]: Because I used cockpit-components-listing.jsx from docker UI, it import the less

**Comment by QiWang19** [comment]
@martinpitt I've added some tests. PTAL.

**Comment by garrett** [comment]
@QiWang19: Hey! I saw your request in IRC. I'm replying here to let you know you were heard. :wink:

If you have design questions or some HTML and/or CSS questions, I'm happy to help. I even know JavaScript pretty OK. However, I don't know React/JSX... outside of modifying it to make HTML/CSS/JS a bit better. 

As your PR looks like a lot of JSX, I probably can't help at this time in this particular PR.

That all said, I've been a fan of Podman (and everything related, like Buildah) for a while and I'm really looking forward to having Podman support in Cockpit!

**Comment by martinpitt** [comment]
Back from PTO, having a look. Tests on master started failing with

    Failed to enable unit: Unit file io.projectatomic.podman.socket does not exist.

as podman apparently renamed the socket. This will also break this PR.

File: .gitignore [hunk 31]
```diff
@@ -13,3 +13,5 @@ test/common/
 test/images/
 *.pot
 POTFILES*
+tmp/*
+/src/test.jsx
```
Comment (martinpitt) [inline .gitignore]: This looks like a debugging leftover. *Don't* add such local files to .gitignore, you are going to forget about them.

File: test/check-application [hunk 32]
```diff
@@ -14,35 +14,46 @@ import testlib
...
-        b.wait_in_text("#images", "busybox:latest")
-        b.wait_in_text("#images", "alpine:latest")
-
-        # no containers
-        self.assertFalse(b.is_present("#containers li"))
-
+        # FIXME: UI needs to listen to change signals and update automatically
         # create a container
-        m.execute("podman run -d --name swamped-crate busybox sleep 1000")
+        # m.execute("podman run -d --name swamped-crate alpine sleep 10000")
```
Comment (martinpitt) [inline test/check-application]: Leftover stuff, should be dropped.

File: test/check-application [hunk 33]
```diff
@@ -14,35 +14,46 @@ import testlib
...
 
-        # FIXME: UI needs to listen to change signals and update automatically
         b.reload()
```
Comment (martinpitt) [inline test/check-application]: This reload should now be dropped as well.

File: test/vm.install [hunk 34]
```diff
@@ -15,3 +15,4 @@ fi
 # grab a few images to play with; tests run offline, so they cannot download images
 podman pull busybox
 podman pull docker.io/alpine
+# podman rm eager_murdock
```
Comment (martinpitt) [inline test/vm.install]: Leftover.

File: test/check-application [hunk 35]
```diff
@@ -14,35 +14,46 @@ import testlib
...
+        b.wait_in_text("#containers-images", "busybox:latest")
+
+        # show image listing toggle
+        b.wait_present('#containers-images tr:contains("busybox:latest")')
+        b.wait_visible('#containers-images tr:contains("busybox:latest")')
+        b.click('#containers-images tbody tr:contains("busybox:latest") td.listing-ct-toggle')
+        b.wait_visible('#containers-images tbody tr:contains("busybox:latest") + tr button.btn-delete')
+
+        # show all containers to interact with stopped ones
+        m.execute("podman run -d --name swamped-crate busybox sh")
```
Comment (martinpitt) [inline test/check-application]: This cannot work, the container will exit immediately. This is why in my initial tests I ran `sleep 10000` in it.

Comment (martinpitt) [reply test/check-application]: Oh, I see, it's supposed to be in the "exited" state, as you switch to "everything".

**Comment by martinpitt** [comment]
Tests succeeded, landing. Thanks!
